### PR TITLE
Ingress controller templates and values

### PR DESCRIPTION
**Pull request by freynca**
Ingress controller templates and values 

https://github.com/projectriff/riff/issues/199

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@freynca Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=projectriff/helm-charts&pullRequestId=11)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=projectriff/helm-charts&pullRequestId=11) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

File: riff/templates/ingress-http-gateway.yaml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+{{- if .Values.ingress.enabled -}}
+apiVersion: extensions/v1beta1
+kind: Ingress
+metadata:
+  name: {{ template "fullname" . }}-ingress-http-gateway
```
Comment (trisberg) [inline riff/templates/ingress-http-gateway.yaml:5]: We should add labels here as well

File: riff/templates/ingress-controller.yaml [hunk 2]
```diff
@@ -0,0 +1,60 @@
+{{- if .Values.ingress.enabled -}}
+apiVersion: extensions/v1beta1
+kind: Deployment
+metadata:
+  name: {{ template "fullname" . }}-ingress-nginx-controller
+  labels:
+    k8s-app: {{ template "name" . }}-ingress-nginx-controller
```
Comment (trisberg) [inline riff/templates/ingress-controller.yaml:7]: The recommended labeling is in the form of:
```
  labels:
    app: {{ template "name" . }}
    component: function-controller
    chart: {{ .Chart.Name }}-{{ .Chart.Version | replace "+" "_" }}
    release: {{ .Release.Name }}
    heritage: {{ .Release.Service }}
```

File: riff/values.yaml [hunk 3]
```diff
@@ -5,6 +5,16 @@ create:
   crd: true
   faas: true
   rbac: false
+ingress:
+  enabled: true
```
Comment (trisberg) [inline riff/values.yaml:9]: prefer to have this default to false

File: riff/templates/ingress-default-backend.yaml [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        livenessProbe:
+          httpGet:
+            path: /healthz
+            port: 8080
+            scheme: HTTP
+          initialDelaySeconds: 30
+          timeoutSeconds: 5
+        ports:
+        - containerPort: 8080
+        resources:
```
Comment (trisberg) [inline riff/templates/ingress-default-backend.yaml:28]: The helm team recommends not including resources by default and leave that up to the user by having this in the values.yaml:

```
resources: {}
  # We usually recommend not to specify default resources and to leave this as a conscious
  # choice for the user. This also increases chances charts run on environments with little
  # resources, such as Minikube. If you do want to specify resources, uncomment the following
  # lines, adjust them as necessary, and remove the curly braces after 'resources:'.
  # limits:
  #  cpu: 100m
  #  memory: 128Mi
  # requests:
  #  cpu: 100m
  #  memory: 128Mi
```

We haven't put this in yet and have so far ignored `resources`. We'll have to add this at some point for all deployments.

File: riff/templates/ingress-controller.yaml (hunk 2 above)
Comment (trisberg) [reply riff/templates/ingress-controller.yaml:7]: This applies to all templates in this PR, see https://docs.helm.sh/chart_best_practices/#standard-labels

File: riff/templates/ingress-controller.yaml [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+spec:
+  replicas: 1
+  template:
+    metadata:
+      labels:
+        k8s-app: {{ template "name" . }}-ingress-nginx-controller
+    spec:
+      terminationGracePeriodSeconds: 60
+      containers:
+      - image: gcr.io/google_containers/nginx-ingress-controller:0.9.0-beta.7
```
Comment (trisberg) [inline riff/templates/ingress-controller.yaml:17]: Should this be in values.yaml for consistency with other deployments?

File: riff/templates/ingress-default-backend.yaml [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+  replicas: 1
+  template:
+    metadata:
+      labels:
+        k8s-app: {{ template "name" . }}-ingress-default-http-backend
+    spec:
+      terminationGracePeriodSeconds: 60
+      containers:
+      - name: {{ .Chart.Name }}-ingress-default-http-backend
+        image: gcr.io/google_containers/defaultbackend:1.0
```
Comment (trisberg) [inline riff/templates/ingress-default-backend.yaml:18]: again, should this be in values.yaml for consistency?

**Review by trisberg — COMMENTED** [review]
I'd like to experiment with using https://hub.kubeapps.com/charts/stable/nginx-ingress as a requirement before we work more on this, might not be suitable, but would be less for us to maintain

**Comment by freynca** [comment]
Using https://hub.kubeapps.com/charts/stable/nginx-ingress with an ingress workload tailored for Riff services works well. In this arrangement, we'd only include the ingress file in our chart (https://github.com/freynca/helm-charts/blob/80fcbbb6e8ab29fedcc137a728c8671290e9f2bd/riff/templates/ingress-http-gateway.yaml). 

Advantage: fewer workload templates to maintain/upgrade
Disadvantage: operator has to install/upgrade two Helm charts

**Comment by trisberg** [comment]
Closing for now, we'll review Ingress solutions at some later time. Thanks for doing the research.

**Pull request closed without merging**